swarm/src/behaviour.rs

```diff
@@ -722,3 +850,364 @@ impl Default for CloseConnection {
         CloseConnection::All
     }
 }
+
+/// Enumeration with the list of the possible events
+/// to pass to [`on_swarm_event`](NetworkBehaviour::on_swarm_event).
+pub enum FromSwarm<'a, Handler: IntoConnectionHandler> {
```

Comment:
Why isn't there `Clone` derive? I need to forward this to underlying protocols and it seems like I need to match on all variants just to be able to clone and wrap it back into `FromSwarm` or am I missing something?

Reply:
There are multiple variants that contain a `ConnectionHandler` which can't be cloned.

See https://github.com/libp2p/rust-libp2p/issues/3046.

Reply:
Sometimes it feels like we're going through a lot of painful contortions just to avoid saying `Arc`.

Reply:
> Sometimes it feels like we're going through a lot of painful contortions just to avoid saying `Arc`.

In this case, we would also have to say `try_unwrap`: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap

Possible but doesn't feel very clean. I don't particularly like both solutions but I think we are getting one step closer with the new event into making composition easier.

https://github.com/libp2p/rust-libp2p/pull/3099 will help with this too because we are removing a few more mentions of the handler there.